src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -326,6 +326,26 @@ private ValueTask<HttpConnection> GetOrReserveHttp11ConnectionAsync(Cancellation
 
             // See if we have an HTTP2 connection
             Http2Connection http2Connection = _http2Connection;
+            if (http2Connection != null)
+            {
+                TimeSpan lifetime = _poolManager.Settings._pooledConnectionLifetime;
+                bool lifetimeExpired = lifetime != Timeout.InfiniteTimeSpan &&
+                                        (lifetime == TimeSpan.Zero || http2Connection.CreationTime + lifetime <= DateTime.UtcNow);
+
+                if (lifetimeExpired)
+                {
+                    _http2Connection = null;
```

Comment:
It seems like there's a race here in setting _http2Connection to null. We're not holding a lock when we do this, so we don't know that the connection hasn't changed out from underneath us.

There's a method in this file called InvalidateHttp2Connection, which I think we probably want to use here. The intent of that method is to indicate that the Http2Connection should no longer be used by the pool, whether because of lifetime or idle or graceful shutdown from the server or connection failure. 

That said, InvalidateHttp2Connection itself may not be entirely correct today.... but that's the intent of it.

Reply:
ok. I was using this function in my earlier code version. I assumed that since we assign _http2Connection to local variable, reset of it may not matter. I'll update the code as you suggested. 

BTW do you think we _should_ track existing connection waiting for completion? With streaming, they can live for a while and maybe MaxConnectionsPerServer may matter if set very low.

Reply:
If I'm understanding the question, I think there are two different cases to consider:

(1) What happens when multiple requests are issued in parallel, and we don't currently have a connection? 

This is complicated by the fact that we don't know in advance if the server will actually support HTTP2. If it does, then it seems to me that we want all pending requests to wait for this connection to be established and then use it. That's what the current code does.

However, this also means that if the server doesn't support HTTP2, then on the first connection attempt, we will block all pending requests until we discover that HTTP2 was not supported. We will then fall back to HTTP/1.1 over SSL and establish as many connections as needed to satisfy all the parallel requests, as we've always done. This does represent a small change from current behavior, but it seems like the most reasonable approach to me. A user can always disable HTTP2 to avoid this.

(2) Would we ever want to actively use more than one HTTP2 connection at the same time, e.g. to get around MaxConnectionsPerServer?

The RFC is pretty clear here. Section 9.1: "Clients SHOULD NOT open more than one HTTP/2 connection to a given host and port pair." I don't see any reason to deviate from this. In particular, if a server sets MaxConnectionsPerServer to a low value, presumably it has a reason for doing so and we should not try to circumvent this.

Reply:
I'm thinking primarily about second case. Lifetime expired on existing connection with existing requests. 
We open new one and than we have two connections now. Does it violate the specification? If so, we would need to delay new requests, right? 
If this is ok, one could see cases when we build up more than expired connection.

Reply:
From the RFC, same section:

```
   A client can create additional connections as replacements, either to
   replace connections that are near to exhausting the available stream
   identifier space (Section 5.1.1), to refresh the keying material for
   a TLS connection, or to replace connections that have encountered
   errors (Section 5.4.1).
```

I think that when we hit the lifetime limit and open a second connection, this counts as a "replacement", broadly speaking.

> If this is ok, one could see cases when we build up more than expired connection.

I disagree. A "replacement" connection is very different than two active connections, in terms of both utility and implementation.

Reply:
I agree this fits the definition of "replacement". 
For one, there will never be new request. 
The point was that there may be more then one connection like this over time. But if we count "active" connections there will always be one or less.